recipes/ncbi-cxx-toolkit-public/26.0/conanfile.py

```diff
@@ -0,0 +1,1532 @@
+    @property
+    def _src(self):
+        return "src"
+
+    def _get_Source(self):
+        self.tk_tmp_tree = tempfile.mkdtemp(dir=os.getcwd())
+        src = os.path.normpath(os.path.join(self.recipe_folder, "..", "source", self._src))
+        src_found = False;
+        if os.path.isdir(src):
+            self.tk_src_tree = src
+            src_found = True;
+        else:
+            tk_url = self.conan_data["sources"][self.version]["url"]
+            print("getting Toolkit sources...")
+            print("from url: " + tk_url)
+            curdir = os.getcwd()
+            os.chdir(self.tk_tmp_tree)
+            tools.get(**self.conan_data["sources"][self.version], strip_root = True)
+            os.chdir(curdir)
+            src_found = True;
+            self.tk_src_tree = self._src
+
+#----------------------------------------------------------------------------
+    def _configure_cmake(self):
+        cmake = CMake(self)
+        cmake.definitions["NCBI_PTBCFG_PACKAGING"] = "TRUE"
+        if self.options.projects != "":
+            cmake.definitions["NCBI_PTBCFG_PROJECT_LIST"] = self.options.projects
+        if self.options.targets != "":
+            cmake.definitions["NCBI_PTBCFG_PROJECT_TARGETS"] = self.options.targets
+        return cmake
+
+#----------------------------------------------------------------------------
+    def validate(self):
+        if self.settings.compiler.get_safe("cppstd"):
+            tools.check_min_cppstd(self, 17)
+        if self.settings.os not in ["Linux", "Macos", "Windows"]:   
+            raise ConanInvalidConfiguration("This operating system is not supported")
+        if self.settings.compiler == "Visual Studio" and str(self.settings.compiler.version) < "15":
+            raise ConanInvalidConfiguration("This version of Visual Studio is not supported")
+        if self.settings.compiler == "Visual Studio" and self.options.shared:
+            runtime = self.settings.compiler.get_safe("runtime")
+            if runtime == "MT" or runtime == "MTd":
+                raise ConanInvalidConfiguration("This configuration is not supported")
+        if self.settings.compiler == "gcc" and str(self.settings.compiler.version) < "7":
+            raise ConanInvalidConfiguration("This version of GCC is not supported")
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+
+#----------------------------------------------------------------------------
+    def requirements(self):
+        if not self.version in self.conan_data["sources"].keys():
+            raise ConanException("Invalid Toolkit version requested. Available: " + ' '.join(self.conan_data["sources"].keys()))
+        self.output.info("Analyzing requirements. Please wait...")
+        curdir = os.getcwd()
+        self._get_Source()
```

Comment:
Shouldn't sources be downloaded in the `source` method 🤔

Reply:
You see, the requirements are calculated dynamically. There is an option to build all libraries (by default) or only some of them. When we build a subset, we do not need all the dependent packages. In order to find out what we really need, we get sources in "requirements" method and analyze. Then, "source" method just moves sources from temporary location into Conan's.